lib/galaxy/tools/__init__.py

```diff
@@ -2510,42 +2510,68 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
         )
 
 
-class FilterFailedDatasetsTool(DatabaseOperationTool):
-    tool_type = 'filter_failed_datasets_collection'
-    require_dataset_ok = False
+class FilterDatasetsTool(DatabaseOperationTool):
+
+    def _get_new_elements(self, history, elements_to_copy):
+        new_elements = odict()
+        for dce in elements_to_copy:
+            element_identifier = dce.element_identifier
+            copied_value = dce.element_object.copy()
+            if getattr(copied_value, "history_content_type", None) == "dataset":
+                history.add_dataset(copied_value, set_hid=False)
+            new_elements[element_identifier] = copied_value
+        return new_elements
 
     def produce_outputs(self, trans, out_data, output_collections, incoming, history, **kwds):
-        hdca = incoming["input"]
+        collection = incoming["input"]
 
-        assert hdca.collection.collection_type == "list" or hdca.collection.collection_type == 'list:paired'
+        if hasattr(collection, 'element_object'):
+            # A list
+            elements = collection.element_object.elements
+            collection_type = collection.element_object.collection_type
+        else:
+            # A pair
+            elements = collection.collection.elements
+            collection_type = collection.collection.collection_type
+        # We only process list or paired collections. Higher order collection will be mapped over
```

Comment:
Should lines 2533 and 2536 be updated to say `list of pair` instead of `pair`/`paired collections` ?